### PR TITLE
Fix: Prevent `ResourceWarning` when using `pytest-xdist`

### PR DESCRIPTION
**Pull request by francis-clairicia**
Hello!
You told me in issue #26 to let you know if I noticed any problems, so here I am.

In an edge case which is very difficult to reproduce, the test suite could end with a small:
```
sys:1: ResourceWarning: unclosed <socket.socket fd=11, family=2, type=1, proto=0, laddr=('127.0.0.1', 33470), raddr=('127.0.0.1', 43621)>
```
And, after investigations, these are the sockets opened by the reporters which are not closed at the end of the process.

### CODE REVIEW

**Comment by str0zzapreti** [comment]
👀

**Comment by str0zzapreti** [comment]
@francis-clairicia Awesome, thanks for catching this. Do you have steps to replicate the issue? I saw you added a test matrix for xdist, but I only have the one xdist test currently in the suite which is specifically for reporting output. It would be ideal to have a test that covers this.

**Comment by francis-clairicia** [comment]
What I do know is this:
- it appears randomly
- it doesn't happen on all Python versions (in 3.11 I see it, in 3.12 I don't)
- not all workers issue this warning
- it meets special conditions (e.g. add a test that catches warnings)

I'll try to reduce the scope to write the test.

**Comment by francis-clairicia** [comment]
@str0zzapreti Hello ! I found a way to test the warning in the main process (with `ReportServer`), but not for the xdist workers.

But at least there is a unit test.

**Comment by str0zzapreti** [comment]
Hey @francis-clairicia thanks for updating the test. I'll take a look as soon as I can

**Comment by pashkatrick** [comment]
any upds?

**Comment by francis-clairicia** [comment]
flake8 issues fixed.

**Comment by str0zzapreti** [comment]
Sorry for the delay @francis-clairicia I still want to test a couple of things but it looks good so far. Been very busy for the last couple months, but I should have some time today.

File: pytest_retry/server.py [hunk 1]
```diff
@@ -41,11 +40,12 @@ def run_server(self) -> None:
...
             conn, _ = self.sock.accept()
 
-            while True:
-                chunk = conn.recv(128)
-                if not chunk:
-                    break
-                self.stream.write(chunk.decode("utf-8"))
+            with conn:
+                while True:
+                    chunk = conn.recv(4096)
```
Comment (str0zzapreti) [inline pytest_retry/server.py:45]: Why the change to the chunk size?

File: tests/test_retry_plugin.py [hunk 2]
```diff
@@ -940,8 +945,8 @@ def test_eventually_passes():
     assert len([line for line in result.outlines if line.startswith('\t  File')]) == len(verbosity)
 
 
-@mark.skipif(xdist_installed is False, reason="Only run if xdist is installed locally")
-def test_xdist_reporting_compatability(testdir):
+@xdist_test_marker
+def test_xdist_reporting_compatibility(testdir):
```
Comment (str0zzapreti) [inline tests/test_retry_plugin.py:949]: Ha, thank you. I think I've fixed this typo twice on different branches only to not merge it

File: tests/test_retry_plugin.py [hunk 3]
```diff
@@ -969,3 +974,54 @@ def test_moar_flaky() -> None:
...
+                func_a_that_randomly_warns()
+
+        def test_flaky_b():
+
+            with pytest.warns(MyWarning):
+                func_b_that_randomly_warns()
+        """
+    )
+
+    # The test MUST be run in a subprocess because the warnings appears
```
Comment (str0zzapreti) [inline tests/test_retry_plugin.py:1022]: Neat trick. Nitpick: just fix the pluralization here (either `warning` or `appear`)

File: tests/test_retry_plugin.py [hunk 4]
```diff
@@ -969,3 +974,54 @@ def test_moar_flaky() -> None:
...
+    # on pytest teardown
+    result = testdir.runpytest_subprocess("-n", "2", "--retries", "3", "-Werror")
```
Comment (str0zzapreti) [inline tests/test_retry_plugin.py:1024]: "-W" "error". Although fwiw, this isn't a necessary param since you're using the `pytest.warn()` context manager, so feel free to just remove it.

File: tox.ini [hunk 5]
```diff
@@ -1,20 +1,21 @@
 [tox]
 minversion = 3.9.0
-envlist = py39, py310, py311, flake8, mypy
+envlist = {py39, py310, py311}-{standard, xdist}, flake8, mypy
```
Comment (str0zzapreti) [inline tox.ini:3]: Running with and without pytest-xdist seems unnecessary here IMO, since the other tests won't invoke it. Just set this to run once with xdist installed.

File: pytest_retry/server.py (hunk 1 above)
Comment (francis-clairicia) [reply pytest_retry/server.py:45]: To reduce the number of `recv()` calls when there is huge tracebacks in verbose mode.

File: tox.ini (hunk 5 above)
Comment (francis-clairicia) [reply tox.ini:3]: I think it would be interesting to see if adding `pytest-xdist` would break certain features.

File: tests/test_retry_plugin.py (hunk 4 above)
Comment (francis-clairicia) [reply tests/test_retry_plugin.py:1024]: The `-W` flag is necessary to force pytest to log the `ResourceWarning` if it appears, because with `-Wdefault`, `ResourceWarning` is ignored. But somehow it gets logged when `pytest.warns()` is used.

**Comment by francis-clairicia** [comment]
It's not a critical issue, but when your library is supposed to be a socket encapsulation and you see this at the end of the test suite, it makes you tick because it's hard to tell at first if it's coming from your code or from pytest. ^^

To be honest, I use your plugin to avoid using rerunfailures. So well, I'm a bit lazy to report this problem on their repo when I haven't come across the problem directly.
What interests me most is that my test suite (which is already far too imposing for what it is) is completed.

**Comment by str0zzapreti** [comment]
Right, totally understand!

File: tox.ini (hunk 5 above)
Comment (francis-clairicia) [reply tox.ini]: I've changed my mind. I removed the matrix.

**Comment by francis-clairicia** [comment]
@str0zzapreti Changes done. Last review?

File: pytest_retry/server.py (hunk 1 above)
Comment (str0zzapreti) [reply pytest_retry/server.py:45]: I suspect that this wouldn't make a meaningful difference (I'm open to seeing benchmarks proving otherwise, however) but I didn't see any issues with this bufsize in my research, so I guess I'm ok with it

File: .github/workflows/tests.yaml [hunk 6]
```diff
@@ -5,6 +5,10 @@ on:
     branches: [ main ]
   pull_request: ~
 
+concurrency:
```
Comment (str0zzapreti) [inline .github/workflows/tests.yaml:8]: I don't really see the need for this. Each test run only takes a few seconds

File: pytest_retry/server.py (hunk 1 above)
Comment (francis-clairicia) [reply pytest_retry/server.py:45]: This isn't a "problem" per se, but tracebacks can quickly reach 1000 characters when the project gets big enough (like mine, for example).

And more generally, when it comes to sockets, the size chosen varies between 4KB and 8KB to receive the maximum amount of data with a small buffer.

(Threads on the xdist workers side send the entire traceback at once anyway).

I can do benchmarks, but it's already been shown by others that it's more efficient to do recv() with larger buffers for data processing. I recommend keeping this modification, it costs nothing.

**Comment by francis-clairicia** [comment]
@str0zzapreti Changes done.

File: pytest_retry/server.py (hunk 1 above)
Comment (str0zzapreti) [reply pytest_retry/server.py:45]: Yeah, like I said I think this change is fine.

**Review by str0zzapreti — APPROVED** [review]
